pimd/pim_upstream.c

```diff
@@ -696,21 +696,47 @@ void pim_upstream_switch(struct pim_instance *pim, struct pim_upstream *up,
 				join_timer_start(up);
 			}
 		}
+		if (old_state != new_state)
+			pim_upstream_update_use_rpt(up, true /*update_mroute*/);
 	} else {
+		bool old_use_rpt;
+		bool new_use_rpt;
+		bool send_xg_jp = false;
 
 		forward_off(up);
 		if (old_state == PIM_UPSTREAM_JOINED)
 			pim_msdp_up_join_state_changed(pim, up);
 
+		if (old_state != new_state) {
+			old_use_rpt =
+				!!PIM_UPSTREAM_FLAG_TEST_USE_RPT(up->flags);
+			pim_upstream_update_use_rpt(up, true /*update_mroute*/);
+			new_use_rpt =
+				!!PIM_UPSTREAM_FLAG_TEST_USE_RPT(up->flags);
+			if (new_use_rpt &&
+					(new_use_rpt != old_use_rpt) &&
+					up->parent)
+				/* we have decided to switch from the SPT back
+				 * to the RPT which means we need to cancel
+				 * any previously sent SGrpt prunes immediately
+				 */
+				send_xg_jp = true;
+		}
+
 		/* IHR, Trigger SGRpt on *,G IIF to prune S,G from RPT towards
 		   RP.
 		   If I am RP for G then send S,G prune to its IIF. */
-		if (pim_upstream_is_sg_rpt(up) && up->parent
-		    && !I_am_RP(pim, up->sg.grp)) {
+		if (pim_upstream_is_sg_rpt(up) && up->parent &&
+				!I_am_RP(pim, up->sg.grp))
+			send_xg_jp = true;
+		else
+			pim_jp_agg_single_upstream_send(&up->rpf, up,
+							0 /* prune */);
+
+		if (send_xg_jp) {
```

Comment:
IMO, sending (*,g) join shall not change SGRpt prune state in rpf neighbor.
We have to send SGRpt Join instead of *,g join.

Reply:
Sending a "compound" (*,G) join to the RPF(RP) neighbor does re-instate sources that are not in the prune list. Please look for "Prune-Pending-Tmp" in RFC7761 for further details.

Reply:
Agreed. (*,G) join with no SGRpt prune in the JP message will put SGRpt state on interface to noInfo state.

But we still have a problem. There can be many SGRpt(S1,G,Rpt) (S2,G,Rpt), etc..
We don't to affect all those by sending just a *,g join alone. 

Sending S,G,Rpt Join will solve the issue precisely. RFC also says that.
RFC 7761 Sec 4.5.7 says when SGRpt prune desired becomes FALSE

      PruneDesired(S,G,rpt)->FALSE
            See macro above.  This transition is only relevant in the
            "Pruned" state.

            If the router is in the Pruned(S,G,rpt) state, and
            PruneDesired(S,G,rpt) changes to FALSE, this could be
            because the router no longer has RPTJoinDesired(G) true, or
            it now wishes to receive traffic from S again.  If it is the
            former, then this transition should not happen, but instead
            the "RPTJoinDesired(G)->FALSE" transition should happen.
            Thus, this transition should be interpreted as
            "PruneDesired(S,G,rpt)->FALSE AND RPTJoinDesired(G)==TRUE".

            The action is to send a Join(S,G,rpt) to RPF'(S,G,rpt).

Reply:
sorry missed your follow up q. compound (*,G) joins always include the pruned source. it is the exact message as the periodic join.

Let's say you start with (*, 239.1.1.1) and source 1.1.1.1, 2.2.2.2 are pruned from the RPT. Now 1.1.1.1 falls back to the RPT. The next compound join will be for (*, 239.1.1.1) with 2.2.2.2 in the pruned source list.